extra/modules/fiftyone-devicedetection/src/main/java/org/prebid/server/hooks/modules/fiftyone/devicedetection/v1/core/SecureHeadersRetriever.java

```diff
@@ -0,0 +1,102 @@
+        if (platform != null) {
+            final String platformName = platform.getBrand();
+            if (StringUtils.isNotBlank(platformName)) {
+                secureHeaders.put("header.Sec-CH-UA-Platform", toHeaderSafe(platformName));
+            }
+
+            final List<String> platformVersions = platform.getVersion();
+            if (CollectionUtils.isNotEmpty(platformVersions)) {
+                final StringBuilder stringBuilder = new StringBuilder();
+                stringBuilder.append('"');
+                appendVersionList(stringBuilder, platformVersions);
+                stringBuilder.append('"');
+                secureHeaders.put("header.Sec-CH-UA-Platform-Version", stringBuilder.toString());
+            }
+        }
+
+        final Integer isMobile = userAgent.getMobile();
+        if (isMobile != null) {
+            secureHeaders.put("header.Sec-CH-UA-Mobile", "?" + isMobile);
+        }
+
+        final String architecture = userAgent.getArchitecture();
+        if (StringUtils.isNotBlank(architecture)) {
+            secureHeaders.put("header.Sec-CH-UA-Arch", toHeaderSafe(architecture));
+        }
+
+        final String bitness = userAgent.getBitness();
+        if (StringUtils.isNotBlank(bitness)) {
+            secureHeaders.put("header.Sec-CH-UA-Bitness", toHeaderSafe(bitness));
+        }
+
+        final String model = userAgent.getModel();
+        if (StringUtils.isNotBlank(model)) {
+            secureHeaders.put("header.Sec-CH-UA-Model", toHeaderSafe(model));
+        }
+
+        return secureHeaders;
+    }
+
+    private static String toHeaderSafe(String rawValue) {
+        return '"' + rawValue.replace("\"", "\\\"") + '"';
+    }
+
+    private static String brandListToString(List<BrandVersion> versions) {
+        final StringBuilder stringBuilder = new StringBuilder();
+        for (BrandVersion nextBrandVersion : versions) {
+            final String brandName = nextBrandVersion.getBrand();
+            if (brandName == null) {
+                continue;
+            }
+            if (!stringBuilder.isEmpty()) {
+                stringBuilder.append(", ");
+            }
+            stringBuilder.append(toHeaderSafe(brandName));
+            stringBuilder.append(";v=\"");
+            appendVersionList(stringBuilder, nextBrandVersion.getVersion());
+            stringBuilder.append('"');
+        }
+        return stringBuilder.toString();
+    }
+
+    private static void appendVersionList(StringBuilder stringBuilder, List<String> versions) {
+        if (CollectionUtils.isEmpty(versions)) {
+            return;
+        }
+        boolean isFirstVersionFragment = true;
+        for (String nextFragment : versions) {
+            if (!isFirstVersionFragment) {
+                stringBuilder.append('.');
+            }
+            stringBuilder.append(nextFragment);
+            isFirstVersionFragment = false;
+        }
+    }
```

Comment:
What about this approach? :p
```java
    private static void appendVersionList(StringBuilder stringBuilder, List<String> versions) {
        if (CollectionUtils.isEmpty(versions)) {
            return;
        }

        stringBuilder.append('.');
        stringBuilder.append(versions.getFirst());
        for (int i = 1; i < versions.size(); i++) {
            stringBuilder.append(versions.get(i));
        }
    }
```

Reply:
:warning: That would invert the logic:

- from "before each segment _except first_"
  - `["111", "0", "5563", "146"]` -> `"111.0.5563.146"`
- to "before _first_ segment _only_"
  - `["111", "0", "5563", "146"]` -> `".11105563146"`

Reply:
Ah, yea, my bad... Tnx for pointing that out!

```java
    private static void appendVersionList(StringBuilder stringBuilder, List<String> versions) {
        if (CollectionUtils.isEmpty(versions)) {
            return;
        }

        stringBuilder.append(versions.getFirst());
        for (int i = 1; i < versions.size(); i++) {
            stringBuilder.append('.');
            stringBuilder.append(versions.get(i));
        }
    }
```

Reply:
✔️ Applied, thanks